### PR TITLE
Fix #5089: Implemented playlist gestures for iOS.

### LINKED ISSUE #5089: Playlist player panning and tapping gestures for media navigation (iPadOS)
## Problem Description/Feature overview

For easier movie navigation on an iPad, add ability to scroll over the media by performing a panning gesture over the playerLayer. Also navigation gestures like double taps on the left/right sides of the playerLayer to skip 15 seconds back and forth would create a better experience.

## Design 

I've implemented these features in my local build. It works like shown on the video:

ProposalVideo

In general, the experience is similar to one of the official YouTube player for iPad. 
It proved to be quite convenient, at least for myself.

### Additional information 

Perhaps some basic player design improvements would be needed to indicate to the user that the feature is available to use.

I would be happy to submit a pull request if Brave/iOS team considers this feature worth adding into the official branch.

### PR DESCRIPTION
**Pull request by yuriymacdev**
Pan the playerLayer to advance the timeline (horizontally) and change the volume (vertically).
Double tap the playerLayer to toggle play/pause (central area) and advance (rightmost area)/retreat (leftmost area) by 15 seconds.

## Summary of Changes

This pull request implemented the feature, described at #5089

## Fix Notes

- Double tap on the video player's view (playerLayer) does not toggle player gravity (toggling between fitting full width or full height fitting of the video) anymore, as the double tapping gesture has been reassigned to other purposes.
I suggest adding the player gravity toggle as a separate UI button in player overlay.
- Occasionally (rarely), for some reason rewinding the video triggers the particle effect (fireworks). Is it a regression or was it happening before?
- The official iOS Youtube app allows rewinding the video by panning the videoplayer's view only after the user holds his finger on the screen for some fraction of a second, thus preventing unwanted timeline movements. I found that this hold is not necessary for me, but may be worth considering adding for the others.
- It may be worth adding some visual cues for the user so he knows these gestures are now available. Perhaps some semi-transparent overlay when the player area is being touched?
- The feature currently works for all iOS devices (not only iPads).

## Test Plan:

- Open a video in a playlist
- Drag (pan) the video player view in horizontal / vertical direction, make sure it rewinds the video (when panned horizontally) and changes the volume (when panned vertically) accordingly.
- Double tap on the left, middle, right area of the player. Make sure it goes back 15 seconds, toggles play/pause, goes forward 15 seconds accordingly.
- Regression testing.

## Screenshots:

screenshot

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 1]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
+            // Retreat by 15 seconds.
+            self.seekBackwards()
+        // If we got a tap in 0.6width...1width (right part of the view).
+        } else if tapLocation.x > 0.6*self.bounds.size.width && tapLocation.x < self.bounds.size.width {
+            // Advance video by 15 seconds.
+            self.seekForwards()
+        // If there's a click in the central area
+        } else if tapLocation.x > 0.4*self.bounds.size.width && 0.4*self.bounds.size.width < 0.6*self.bounds.size.width {
+            // Pause/resume the video
+            if delegate?.isPlaying ?? false {
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]: `if delegate?.isPlaying == true`

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 2]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
     @objc
     private func onOverlayDoubleTapped(_ gestureRecognizer: UITapGestureRecognizer) {
-        delegate?.togglePlayerGravity(self)
+        // Originally double tap changed size of the video to fit the screen,
+        // now we will use this gesture to advance/go back by 15 secs.
+        // Perhaps we should add a button to the player's toolbar overlay for toggling gravity.
+        // delegate?.togglePlayerGravity(self)
+
+        // Advance or go back by 15 seconds
+        let tapLocation: CGPoint = gestureRecognizer.location(in: self)
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]: `let tapLocation = gestureRecognizer.location(in: self)`

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 3]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
+
+        // If we got a tap in 0width...0.4width (left part of the view).
+        if tapLocation.x > 0 && tapLocation.x < 0.4*self.bounds.size.width {
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]: Space between `0.4 * self.bounds.width` and also no need for `bounds.size.width`. Can just do `bounds.width`.

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 4]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
+            // Retreat by 15 seconds.
+            self.seekBackwards()
+        // If we got a tap in 0.6width...1width (right part of the view).
+        } else if tapLocation.x > 0.6*self.bounds.size.width && tapLocation.x < self.bounds.size.width {
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]: Same thing here :)

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 5]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
+                let currentSeconds = CGFloat(currentPlayer.currentTime().value) / CGFloat((currentPlayer.currentTime().timescale))
+                dragStartedTimelinePos = currentSeconds
+            }
+        }
+
+        super.touchesBegan(touches, with: event)
+    }
+
+    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
+        if event?.type == UIEvent.EventType.touches {
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift:269]: `if event?.type == .touches` should work.

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 6]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
+                        currentDraggingDirection = .verticalDirection
+                    } else {
+                        currentDraggingDirection = .horizontalDirection
+                    }
+                }
+
+                if currentDraggingDirection == .verticalDirection {
+
+                    // Vertical dragging adjusts volume.
+                    let vDragRatio = 4*gestureRecognizer.translation(in: self).y/self.bounds.height // (-0.5 ... 0.5)*4
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]: space between operators `*` and `/`

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 7]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
+                    
+                    if vDragRatio != 0 && self.dragStartedVolume >= 0 {
+                        if let slider = volumeView.subviews.first(where: { $0 is UISlider }) as? UISlider {
+                            var finalVal: Float = 0
+                            finalVal = self.dragStartedVolume + Float(-vDragRatio)
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]: `let finalVal = self.dragStartedVolume -vDragRatio`

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 8]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
+
+                            // Making sure finalVal is within 0..1 interval
+                            if finalVal < 0 {
+                                finalVal = 0
+                            } else if finalVal > 1 {
+                                finalVal = 1
+                            }
+                            slider.value = finalVal
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]: `self.value = min(max(0.0, finalVal), 1.0)`

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift [hunk 9]
```diff
@@ -202,9 +223,160 @@ class VideoView: UIView, VideoTrackerBarDelegate {
...
+                    }
+                    var finalSeconds = dragStartedTimelinePos + dragRatio*vidDurationSeconds
+
+                    // Making sure we're not out of bounds
+                    if finalSeconds<0 {
+                        finalSeconds = 0
+                    }
+                    if finalSeconds > vidDurationSeconds {
+                        finalSeconds = vidDurationSeconds
+                    }
```
Comment (Brandon-T) [inline Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]:
```swift
var finalSeconds = dragStartedTimelinePos + dragRatio * vidDurationSeconds
finalSeconds = min(max(0.0, finalSeconds), vidDurationSeconds)
```

**Review by Brandon-T — CHANGES_REQUESTED** [review]

**Comment by Brandon-T** [comment]
Left a few comments just for code-style. Otherwise the code looks great. Thanks! :)

**Comment by yuriymacdev** [comment]
@Brandon-T, should be fixed. Please have a look.

File: Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift (hunk 7 above)
Comment (yuriymacdev) [reply Client/Frontend/Browser/Playlist/VideoPlayer/UI/VideoPlayer.swift]: I think here we do need explicit type conversion as one operand is CGFloat and the other one is Float.

**Review by Brandon-T — APPROVED** [review]

**Comment by bsclifton** [comment]
Nice work, @yuriymacdev! 😄👍

**Pull request merged**